cfme/services/requests.py

```diff
@@ -1,49 +1,361 @@
+        self.partial_check = partial_check
+        self.view = navigate_to(self, 'All')
+        self.row = self.view.find_request(self.cells, self.partial_check)
+        self.request_id = self.row.request_id.text
+        self.description = self.row.description.text
+
+    @property
+    def exists(self):
+        return True if self.view.find_request(cells=self.cells, partial_check=self.partial_check)\
+            else False
+
+    def reload(self):
+        self.view.reload.click()
+
+    def update_row(self):
+        """Updates Request object details - last message, status etc
+        """
+        self.reload()
+        self.row = self.view.find_request(self.cells, self.partial_check)
+
+    def approve_request(self, reason, cancel=False):
+        """Approves request with specified reason
+        Args:
+            reason: Reason for approving the request.
+            cancel: Whether to cancel the approval.
+        """
+        view = navigate_to(self, 'Details')
+        view.approve(reason, cancel)
+        self.update_row()
+
+    def deny_request(self, reason, cancel=False):
+        """Open the specified request and deny it.
+        Args:
+            reason: Reason for denying the request.
+            cancel: Whether to cancel the denial.
+        """
+        view = navigate_to(self, 'Details')
+        view.deny(reason, cancel)
+        self.update_row()
+
+    def delete_request(self, cancel=False):
+        """Open the specified request and delete it.
+        Args:
+            cancel: Whether to cancel the deletion.
+        """
+        view = navigate_to(self, 'Details')
+        view.delete(cancel)
+
+    def is_finished(self):
+        """Helper function checks if a request is completed
+        """
+        self.update_row()
+        # self.row = self.view.find_request(self.cells, self.partial_check)
+        if self.row.request_state.text in self.REQUEST_FINISHED_STATES:
+            return True
+        else:
+            return False
+
+    def if_succeeded(self):
+        self.row = self.view.find_request(self.cells, self.partial_check)
+        if self.is_finished() and self.row.status.text == 'Ok':
+            return True
+        else:
+            return False
+
+    def debug_request(self):
+        self.row = self.view.find_request(self.cells, self.partial_check)
+        logger.debug('Last Message of request: {}'.format(self.row.last_message.text))
+
+    @contextmanager
+    def copy_request(self):
+        """Context manager that opens the request for editing and saves or cancels depending on success.
+        """
+        view = navigate_to(self, 'Details')
+        view.toolbar.copy.click()
+        try:
+            yield
+        except Exception:
+            view.cancel.click()
+            raise
+        else:
+            from cfme.provisioning import provisioning_form
+            btn = provisioning_form.submit_copy_button
+            sel.wait_for_element(btn)
+            sel.click(btn)
```

Comment:
decided not to cherry-pick Mike's PR here too - I'll remove this context manager. edit_request will return new Request class object and will fill all forms on it's side